### ISSUE TITLE
NRE in ReviewCodeForInformationDisclosureVulnerabilities

### ISSUE THREAD

**Issue by ashishakalra** [dotnet/roslyn-analyzers#2383]
Analyzer package: Microsoft.NetCore.Analyzers

Package Version:  v2.9.2

Diagnostic ID:
AD0001	Analyzer 'Microsoft.NetCore.Analyzers.Security.ReviewCodeForInformationDisclosureVulnerabilities' threw an exception of type 'System.NullReferenceException' with message 'Object reference not set to an instance of an object.'.

Stack:
Analyzer 'Microsoft.NetCore.Analyzers.Security.ReviewCodeForInformationDisclosureVulnerabilities' threw the following exception:
'Exception occurred with following context:
Compilation: <redacted>
ISymbol: Logger_Exceptions_ShouldSerializeCorrectly (Method)

System.NullReferenceException: Object reference not set to an instance of an object.
   at Analyzer.Utilities.Extensions.IEnumerableExtensions.Dispose[T](IEnumerable`1 collection)
   at Microsoft.CodeAnalysis.FlowAnalysis.DataFlow.DataFlowAnalysisResultBuilder`1.Dispose()
   at Microsoft.CodeAnalysis.FlowAnalysis.DataFlow.DataFlowAnalysis`5.Run(TAnalysisContext analysisContext)
   at Microsoft.CodeAnalysis.FlowAnalysis.DataFlow.DataFlowAnalysis`5.<>c__DisplayClass8_0.<GetOrComputeResultCore>b__0(DataFlowOperationVisitor`4 _)
   at Analyzer.Utilities.SingleThreadedConcurrentDictionary`2.GetOrAdd(TKey key, Func`2 valueFactory)
   at Microsoft.CodeAnalysis.FlowAnalysis.DataFlow.DataFlowAnalysis`5.GetOrComputeResultCore(TAnalysisContext analysisContext, Boolean cacheResult)
   at Analyzer.Utilities.FlowAnalysis.Analysis.TaintedDataAnalysis.TaintedDataAnalysis.GetOrComputeResultForAnalysisContext(TaintedDataAnalysisContext analysisContext)
   at Analyzer.Utilities.FlowAnalysis.Analysis.TaintedDataAnalysis.TaintedDataAnalysis.GetOrComputeResult(ControlFlowGraph cfg, Compilation compilation, ISymbol containingMethod, TaintedDataSymbolMap`1 taintedSourceInfos, TaintedDataSymbolMap`1 taintedSanitizerInfos, TaintedDataSymbolMap`1 taintedSinkInfos, InterproceduralAnalysisConfiguration interproceduralAnalysisConfig)
   at Microsoft.NetCore.Analyzers.Security.SourceTriggeredTaintedDataAnalyzerBase.<>c__DisplayClass6_1.<Initialize>b__4(OperationBlockAnalysisContext operationBlockAnalysisContext)
   at Microsoft.CodeAnalysis.Diagnostics.AnalyzerExecutor.<>c__49`2.<ExecuteBlockActions>b__49_1(ValueTuple`2 data)
   at Microsoft.CodeAnalysis.Diagnostics.AnalyzerExecutor.ExecuteAndCatchIfThrows_NoLock[TArg](DiagnosticAnalyzer analyzer, Action`1 analyze, TArg argument, Nullable`1 info)
-----
'.

Repro steps:
Don't have isolated steps.

Expected behavior: No Exception.

Actual behavior: NRE

**Comment by dotpaul** [comment]
Hey @mavasani, not sure if this is same root cause as this issue, but here's one case that fails in ReviewCodeForInformationDisclosureVulnerabilitiesTests.cs:

```csharp
        [Fact]
        public void TryUsingTryUsingTry()
        {
            this.VerifyCSharp(@"
using System;
using System.IO;
using System.Threading.Tasks;
using System.Web.UI.WebControls;

public class Class
{
    public BulletedList BulletedList;

    public async Task DoSomethingNotReallyAsync(Stream stream)
    {
        try
        {
            using (stream)
            {
                try
                {
                    using (StreamWriter sw = new StreamWriter(stream))
                    {
                        ValidateStreamIsNotMemoryStream(stream);
                        sw.Write(""Hello world!"");
                    }
                }
                catch (Exception ex)
                {
                    Console.Write(ex.Message);
                }
            }
        }
        catch (Exception e)
        {
            Console.WriteLine(e.Message);
        }
        finally
        {
            
        }
    }

    private static void ValidateStreamIsNotMemoryStream(Stream stream)
    {
        if (stream is MemoryStream)
        {
            throw new ArgumentException(nameof(stream));
        }
    }
}
");
        }
```

**Comment by mavasani** [comment]
@dotpaul Regarding your failing test: I believe the fix needs to go in here: https://github.com/dotnet/roslyn-analyzers/blob/master/src/Utilities/FlowAnalysis/FlowAnalysis/Analysis/TaintedDataAnalysis/TaintedDataAnalysis.TaintedDataOperationVisitor.cs#L77.

`HasAnyAbstractValue(data)` should be implemented as `return data.HasAnyAbstractValue;` instead of `return this.CurrentAnalysisData.HasAnyAbstractValue;`.

**Comment by mavasani** [comment]
@dotpaul I presume the NRE callstack above will be already avoided by your earlier PR: https://github.com/dotnet/roslyn-analyzers/pull/2384. So we probably close this bug once your failing test has been fixed?

**Comment by dotpaul** [comment]
Yep, will close if I don't run into any more issues.  Thanks for debugging that!

**Comment by dotpaul** [comment]
Hi @ashishakalra, would you mind trying the package at https://dotnet.myget.org/feed/roslyn-analyzers/package/nuget/Microsoft.CodeAnalysis.FxCopAnalyzers/2.9.3-beta1.19226.1+513b9bc4, and seeing if that resolves the issue, or if you get a different error, please?

**Comment by ashishakalra** [comment]
Hi @dotpaul, I am getting these errors now with this new package:

Error	IDE1003	Analyzer assembly 'C:\path\.nuget\packages\microsoft.codequality.analyzers\3.0.0-beta2.19218.6\analyzers\dotnet\cs\Microsoft.CodeQuality.Analyzers.dll' depends on 'netstandard, Version=2.0.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51' but it was not found. Analyzers may not run correctly unless the missing assembly is added as an analyzer reference as well.

**Comment by dotpaul** [comment]
Hey @ashishakalra, from that path it sounds like you have version 3.0.0-beta2.19218.6.  Can you try version 2.9.3-beta1.19226.1?  You can get it from here: https://dotnet.myget.org/feed/roslyn-analyzers/package/nuget/Microsoft.CodeAnalysis.FxCopAnalyzers/2.9.3-beta1.19226.1+513b9bc4

**Comment by ashishakalra** [comment]
oops! my bad. Thanks @dotpaul and I can confirm that the issue I reported is fixed with this package you mentioned.
Thanks to you guys for quick turnaround on this.

**Comment by dotpaul** [comment]
Thank you @ashishakalra for reporting and testing the beta!

Fixed by #2384 and #2391.
